src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,370 @@
+from __future__ import annotations
+
+import logging
+from typing import Any, Optional, Sequence, Set
+
+import sentry_sdk
+from celery.exceptions import SoftTimeLimitExceeded
+
+from sentry import options
+from sentry.api.utils import get_date_range_from_params
+from sentry.models.dashboard_widget import (
+    DashboardWidgetQuery,
+    DashboardWidgetQueryOnDemand,
+    DashboardWidgetTypes,
+)
+from sentry.models.organization import Organization
+from sentry.models.project import Project
+from sentry.relay.config.metric_extraction import (
+    HashedMetricSpec,
+    convert_widget_query_to_metric,
+    on_demand_metrics_feature_flags,
+)
+from sentry.search.events import fields
+from sentry.search.events.builder import QueryBuilder
+from sentry.search.events.types import EventsResponse, QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.query import RangeQuerySetWrapper
+
+logger = logging.getLogger("sentry.tasks.on_demand_metrics")
+
+OnDemandExtractionState = DashboardWidgetQueryOnDemand.OnDemandExtractionState
+
+# TTL for cardinality check
+_WIDGET_QUERY_CARDINALITY_TTL = 3600 * 24 * 7  # Cardinality outcome is valid for 7 days.
+
+
+def _get_widget_processing_batch_key() -> str:
+    return "on-demand-metrics:widgets:currently-processing-batch"
+
+
+def _get_widget_query_cardinality_cache_key(widget_query: DashboardWidgetQuery) -> str:
+    return f"check-widget-query-cardinality:{widget_query.id}"
+
+
+def _set_currently_processing_batch(current_batch: int) -> None:
+    cache.set(_get_widget_processing_batch_key(), current_batch, timeout=3600)
+
+
+def _set_cardinality_cache(cache_key: str, is_low_cardinality: bool) -> None:
```

Comment:
It seems like we mock `_set_cardinality_cache` and `_get_previous_processing_batch` in the tests.
We should have an integration test for these if we want to be sure (or not mock them).

Reply:
I'll spy the set cache instead, the value doesn't get re-used anyway so there isn't a chance for it breaking isolation. For previous batch I'll maybe write another small test.